pkg/pillar/zedbox/zedbox.go

```diff
@@ -110,14 +113,151 @@ var (
+		},
+		logger, log)
+
+	subGlobalConfig, err := ps.NewSubscription(pubsub.SubscriptionOptions{
+		AgentName:   "zedagent",
+		MyAgentName: agentName,
+		TopicImpl:   types.ConfigItemValueMap{},
+		Persistent:  false,
+		Activate:    false,
+		WarningTime: warningTime,
+		ErrorTime:   errorTime,
+	})
+	if err != nil {
+		log.Warnf("initSubGlobalConfig(): can't subscribe on global config: %v", err)
+		return
+	}
+	// Wait for initial GlobalConfig
+	if err = subGlobalConfig.Activate(); err != nil {
+		log.Warnf("initSubGlobalConfig(): activate failed: %v", err)
+		return
+	}
+	for {
+		select {
+		case change := <-subGlobalConfig.MsgChan():
+			subGlobalConfig.ProcessChange(change)
+			m, err := subGlobalConfig.Get("global")
+			if err == nil {
+				gc := m.(types.ConfigItemValueMap)
+				gogcForcedLock.Lock()
+				gogcForcedPeriodInSec =
+					gc.GlobalValueInt(types.GOGCForcedPeriodInSec)
+				gogcForcedGainedMemInMiB =
+					gc.GlobalValueInt(types.GOGCForcedGainedMemInMiB)
+				gogcForcedGainedMemPerc =
+					gc.GlobalValueInt(types.GOGCForcedGainedMemPerc)
+				gogcForcedLock.Unlock()
+			}
+		}
+	}
+}
+
+// Listens to main /eve cgroup in hierarchy mode (events always propagate
+// up to the root) and call Go garbage collector with reasonable period
+// when certain amount of memory has been allocated (presumably there is
+// something to free) by an application.
+func readMemoryPressureEvents() {
+	controller, err := cgroups.Load(cgroups.V1, cgroups.StaticPath(""))
+	if err != nil {
+		log.Warnf("readMemoryPressureEvents(): failed to find root cgroups directory")
+		return
+	}
+	defer controller.Delete()
+	event := cgroups.MemoryPressureEvent(cgroups.MediumPressure,
+		cgroups.HierarchyMode)
+	efd, err := controller.RegisterMemoryEvent(event)
+
+	fd := os.NewFile(efd, "efd")
+	defer fd.Close()
+
+	buf := make([]byte, 8)
+
+	var before, after runtime.MemStats
+	var expected, freed uint64
+	var ts time.Time
+
+	// Infinite loop until error or death
+	for {
+		if _, err := fd.Read(buf); err != nil {
+			log.Warnf("readMemoryPressureEvents(): can't read eventfd, exiting loop")
+			return
+		}
+		// GC is called explicitly no more than once every @period, and
+		// only if the application has already gained at least
+		// @minGainedFatAbs and certain fraction from the last reclaim, so
+		// shortly:
+		//     limit = MAX(minGainedFatAbs, freed * minGainedFatPerc / 100)
+		period, minGainedFatAbs, minGainedFatPerc := getForcedGOGCParams()
+
+		if period == 0 || time.Now().Sub(ts) < period {
+			// Don't call GC frequently in case of many sequential events
+			continue
+		}
+		runtime.ReadMemStats(&before)
+		if before.Alloc < expected {
+			// Not enough allocated since last GC call, skip this event
+			continue
+		}
+		runtime.GC()
+		runtime.ReadMemStats(&after)
+
+		// Careful, unlikely but can be negative if nothing was
+		// reclaimed but allocation has happened just in between the
+		// GC call and stats collection
+		freed = 0
```

Comment:
```suggestion
	var before, after runtime.MemStats
	var expected uint64
	var ts time.Time

	// Infinite loop until error or death
	for {
		if _, err := fd.Read(buf); err != nil {
			log.Warnf("readMemoryPressureEvents(): can't read eventfd, exiting loop")
			return
		}
		// GC is called explicitly no more than once every @period, and
		// only if the application has already gained at least
		// @minGainedFatAbs and certain fraction from the last reclaim, so
		// shortly:
		//     limit = MAX(minGainedFatAbs, freed * minGainedFatPerc / 100)
		period, minGainedFatAbs, minGainedFatPerc := getForcedGOGCParams()

		if period == 0 || time.Now().Sub(ts) < period {
			// Don't call GC frequently in case of many sequential events
			continue
		}
		runtime.ReadMemStats(&before)
		if before.Alloc < expected {
			// Not enough allocated since last GC call, skip this event
			continue
		}
		runtime.GC()
		runtime.ReadMemStats(&after)

		// Careful, unlikely but can be negative if nothing was
		// reclaimed but allocation has happened just in between the
		// GC call and stats collection
		freed := uint64(0)
```